src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Custom/Serializer/Writer.cs

```diff
@@ -0,0 +1,332 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System.Buffers.Binary;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.Custom.Serializer;
+
+internal class Writer
+{
+    private const uint Uint128 = 128;
+    private const ulong Ulong128 = 128;
+    private const int Fixed64Size = 8;
+
+    internal static Encoding Utf8Encoding => Encoding.UTF8;
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static int WriteStringWithTag(ref byte[] buffer, int cursor, int fieldNumber, string value)
+    {
+        return WriteStringWithTag(ref buffer, cursor, fieldNumber, value.AsSpan());
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static int WriteStringWithTag(
+        ref byte[] buffer,
+        int cursor,
+        int fieldNumber,
+        ReadOnlySpan<char> value,
```

Comment:
@vishweshbankwar I pushed an update so `Writer` understands how to write a `ReadOnlySpan<char>` to the buffer. Previously we called `ToString` in `TagWriter` which would have caused a lot of extra allocations. Also updated this so it calls into the code to grow the buffer if needed before writing instead of having two branches with one performing that check per-byte/char written. Should be much faster this way and I think simpler.